Mediator

john-breton opened the pull request:
Hey all, still not feeling really good at all. I tried my best to stitch together most of everything to meet the requirements of demo #1. I think everything we are required to show exists, but obviously, there are still lots of rough patches. I'm not sure if I'm going to work tomorrow but if I do I'll try doing a bit more cleanup in the morning.

john-breton committed:
Add early Mediator

- Added an initial mediator implementation (interface + concrete class).
- Made the guess that we will be keeping a reference to created RUs within the mediator itself.
- Made the mediator a Singleton to ensure data-consistency if multiple admins make use of the system at one time.

john-breton committed:
Correct displayCarrier

- Display carrier will now pull the list of carriers associated with each RU and check to see if the carrier with the passed id exists. If it does, the mediator instructs the carrier to print its information to the console.

john-breton committed:
Cleaned up lambdas

john-breton committed:
Fix incorrect method call

john-breton committed:
Add UI relevent accessible methods

john-breton committed:
Remove references to IPs

cheeryr committed:
UserInterface for demo#1

cheeryr committed:
fix errors

- use CarrierManagementSystemDirector to create carrier.

john-breton committed:
Add carrier creation method handler to mediator

john-breton committed:
Have carriers displayed by RU name

john-breton committed:
Fix typo in method name

cheeryr committed:
change menu options

Add corresponding steps to support menu options

cheeryr committed:
use enum type

cheeryr committed:
remove unused import

john-breton committed:
Merge pull request #2 from RPahlevan/UserInterface

User interface

john-breton committed:
Merge branch 'main' into mediator

john-breton committed:
Merge branch 'main' into mediator

john-breton committed:
Oh yeah. It's all coming together.

- Managed to somehow get everything to stop throwing errors and begin working together.
- Added a Vendor enum to the common package (not needed for demo 1).
- Added a constructor to the temporary RadioUnit class.
- Still lots to be done.

john-breton committed:
Added AtomicBooleans for lambda functions

- Used AtomicBooleans for various lambda functions in ConcreteMediator to ensure only the correct error message is printed.

john-breton committed:
Cleanup

- Made the mediator work nicer with the user interface.
- Made the mediator track created carriers such that a carrier can be created and then later added to an RU after the fact.
- Added stubs for RU related carrier activities.
- Addressed some feedback about the UI.

john-breton committed:
Remove option 11 for now

john-breton committed:
Merge branch 'main' into mediator

john-breton committed:
Adapt changes from latest CarrierManagement merge

john-breton committed:
Further cleanup

- Moved the AlarmStatusLevel enum out of the common to the radiounit package (for now).
- Converted some switches to enhanced switches (thanks Java 14).
- Corrected some dangling Javadoc issues.
- Fixed a typo in one of the mediator method names.
- Added a verbose [DEBUG] tag to all current DemoOneRadioUnit printouts.

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * @param radioUnit The RU that will be registered with the mediator.
+     */
+    private synchronized void register(ManagedRadioUnit radioUnit) {
+        if (!radioUnits.contains(radioUnit)) {
```
I suspect we may want to validate that IP addresses are unique here - if we somehow create two separate `ManagedRadioUnit`s with the same IP address, we should probably either swap them here (unregister existing, register new) or be throwing some sort of error. Maybe this should really be enforced where we build the radios, but something we should be considering here

john-breton replied:
I agree. I think it makes sense for the mediator to do the work here to further leverage the benefits of using the pattern in the first place.

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * @param name The name associated with the targeted RU.
+     */
+    @Override
+    public void printRatType(String name) {
```
I think we should be using IpAddress instead of name - IPs are required to be unique, however we (in theory) could create multiple RUs with the same name

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * @param name The name associated with the targeted RU.
+     */
+    @Override
+    public void printVendor(String name) {
```
Same as above

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * @param name The name associated with the targeted RU.
+     */
+    @Override
+    public void printAlarmStatus(String name) {
```
Same

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * @param ratType The RAT type for the RU.
+     */
+    @Override
+    public synchronized void createRu(String name, Vendor vendor, RatType ratType) {
```
We'll need to add IpAddress somewhere here.

john-breton replied:
Indeed, although for the first demo we had decided to omit it since it technically wasn't a part of the requirement spec. Realistically instead of specifying an RU name in the UI, we should be specifying an IP (since it is valid for two RUs to share the same name according to the requirements). But still needs to be addressed for demo 2.

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+        Carrier carrier;
+        for (ManagedRadioUnit ru : radioUnits) {
+            if (ru.getRadioUnitName().equals(name)) {
+                switch (rfPorts.size()) {
```
Instead of relying on the number of RF ports passed, should we not be getting the RAT type from the radio unit? In the future, it's possible more RAT types could be added that also use either 2 or 4 RF ports.

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     */
+    @Override
+    public synchronized void createCarrier(List<RfPort> rfPorts, FrequencyBand carrierFrequencies, Double transmittingPower) {
+        switch (rfPorts.size()) {
```
Similar comment as above, however since we don't have an RU to get the RAT type from, the FrequencyBand should give us the appropriate information (frequency bands should be specific to RAT types, so we shouldn't ever have another RAT type using the same frequency)

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+                flag.set(true);
+                if (ru.getCarriers().size() == 0) {
+                    System.out.printf(
+                            "[ERROR] The RU with the name %s has no carriers associated with it.%n", name);
```
I'm not convinced this is necessarily an error condition - maybe we just print the number of carriers (regardless of if there are any there), then list the ones there.
Output might look something like:
```
Number of  carriers: 0
```
or
```
Number of carriers: 1
<carrier details>
```

dLeung09 commented on src/main/java/mediator/ConcreteMediator.java:
```diff
@@ -0,0 +1,287 @@
+     * not need to be carriers associated with an RU.
+     */
+    @Override
+    public void printCreatedCarriers() {
```
I'm not entirely sure which requirement this is satisfying - my understanding is that a carrier must always be associated with an RU (from a real life point of view, they don't exist; from a requirements point of view, why else would they have an RF port?), so holding a second list of carriers here might not be necessary.
Printing all carriers could then be achieved by iterating over all the RUs, and listing the carriers setup on those.

john-breton replied:
So I was iffy on this as well. But I think I must have just misunderstood 1.8.3 while I was reviewing it over the weekend because I am in agreement, a carrier will only ever be created and then immediately associated with an RU. There's no need to have it in a limbo state where it could be associated with an RU after the fact (at least according to what Ista has written). So I'll revert this and remove the carrier list since it really won't be needed.

dLeung09 commented on src/main/java/radiounit/AlarmStatusLevel.java:
```diff
@@ -1,7 +1,4 @@
- * 
- */
-package common;
+package radiounit;
```
This should stay in common - it'll be used elsewhere other than just in the radio unit

john-breton replied:
Sounds good, I'll move it back. I wasn't positive it was going to be used by anything else.

dLeung09 commented on src/main/java/radiounit/DemoOneRadioUnit.java:
```diff
@@ -1,154 +1,170 @@
+        System.out.println("[DEBUG] Remove carrier");
+        carriers.forEach(carrier -> {
+            if (carrier.getCarrierId() == carrierId) {
+                //carriers.remove(carrier);
```
We should uncomment this

john-breton replied:
The comments can be removed, I left it like this since we didn't actually make use of the functionality in demo 1 at all, since it wasn't a part of the requirements we were given. These were just some functions I threw together quickly while trying to have a RU work with the UI.

dLeung09 commented on src/main/java/userinterface/UserInterface.java:
```diff
@@ -1,7 +1,496 @@
+     */
+    private static Vendor chooseVendor(Scanner input) {
+        Vendor vendor = null;
+        System.out.println("Please choose a Vendor for Radio Unit:\n"
```
I think a future improvement we could add is to leverage the enum here - there should be a way to list all of the enum values so that adding new vendors won't require a change here. Similarly below, when we parse the user's input to a vendor enum, we could probably add some method to the enum class that takes the id (e.g. number) and returns the appropriate enum.

john-breton replied:
I agree this would make more sense and aim to remove a lot of the code for these methods currently in the UserInterface class, I just didn't have enough time to get around to making that change. But I'll adapt to this in a new commit.

dLeung09 commented on src/main/java/userinterface/UserInterface.java:
```diff
@@ -1,7 +1,496 @@
+     */
+    private static RatType chooseRatType(Scanner input) {
+        RatType rat = null;
+        System.out.println("\nPlease choose a RAT type:\n"
```
Same as above

dLeung09 reviewed (commented):
Not sure if this PR is still relevant (since we already submitted our code to Ista - thanks guys!), but I reviewed it regardless, since I think the code changes here are still relevant and should be merged at some point.
Overall, this looks good though!

john-breton committed:
Merge branch 'main' into mediator

john-breton committed:
Address code-review comments.

THINGS THAT NEED TO BE DONE
 - Changing RU lookup to be done via an IP rather than a name.
THINGS THAT HAVE BEEN DONE
- Moved AlarmStatusLevel back into the common package.
- Created a FrequencyBand interface that the WCDMA and LTE enums now implement. This is meant as a functional interface marker and aims to ensure we don't have random numbers to tell us where to start printing enums values if they are requested at any point during execution.
- Added an IP to the RUs that is randomly assigned on creation. There is also a check to ensure no other RUs currently registered with the system share the IP the newly created RU.
- Removed some menu options that didn't need to exist (specifically, there is no longer a way to create an unassociated carrier).
- Removed the carrier list from the mediator, since it didn't need to exist per the requirements.
- Made the UserInterface take advantage of the enums a bit more to remove about 100+ lines of print statements. The way things function now is that the entire enum is printed (where needed) in a formatted string as output. The values come from the enum itself and not the UserInterface (which was the case before).
- Added the ability to get an RU by name, this is to prevent asking for a RAT Type during carrier setup on an RU that has previously been created (since the RU was already assigned a type, and we can't mix/match RF ports).
- RF Ports will now get the RAT type to determine how many RF ports for a new carrier.
- Made some stuff final where it made sense to do so.
- Added the ability to set a frequency band on a previously created carrier (apart of the requirement spec).

dLeung09 commented on src/main/java/common/LteFrequencyBand.java:
```diff
@@ -0,0 +1,26 @@
+package common;
+
+public enum LteFrequencyBand implements FrequencyBand {
```
Really like this!

dLeung09 reviewed (approved):
https://media.giphy.com/media/cPkUnVPxfmkMWzZfzx/giphy.gif
